### PR TITLE
Ping

### PR DESCRIPTION
**Pull request by mathieufrh**
I try to add a simple and manual clickable icon (choosed very carefully from font awesome...) which allow the user to request a host ping.
I added a column to the host table view (in fact this column contains more buttons, coming from different features I add). I'm not a good web dev so maybe there is a better way to display and handle this icon.
Upon click it call `interface.pingHost` which, in the backend, call `Utils::ping`. This function simply ping a host and return the pong delay, it do not works for Ipv6 but I think there is little work to have it handle IPv6.

### CODE REVIEW

File: src/Utils.cpp [hunk 1]
```diff
@@ -1569,3 +1569,165 @@ void Utils::xor_encdec(u_char *data, int data_len, u_char *key) {
     if(key[y] == 0) y = 0;
   }
 }
+
+/* ****************************************************** */
+
+#define DEFDATALEN (64-ICMP_MINLEN) /* default data length */
```
Comment (simonemainardi) [inline src/Utils.cpp]: defines go in `includes/ntop_defines.h`

File: src/Utils.cpp [hunk 2]
```diff
@@ -1569,3 +1569,165 @@ void Utils::xor_encdec(u_char *data, int data_len, u_char *key) {
...
+		hostname = hnamebuf;
+	}
+	packlen = datalen + MAXIPLEN + MAXICMPLEN;
+	if((packet = (u_char *)malloc((u_int)packlen)) == NULL){
+		ntop->getTrace()->traceEvent(TRACE_ERROR, "malloc error\n");
+		return -1;
+	}
+
+	if((s = socket(AF_INET, SOCK_RAW, IPPROTO_ICMP)) < 0){
+		perror("socket");   /* probably not running as superuser */
```
Comment (simonemainardi) [inline src/Utils.cpp:1645]: this is not the way ntopng handles errors

File: src/Utils.cpp [hunk 3]
```diff
@@ -1569,3 +1569,165 @@ void Utils::xor_encdec(u_char *data, int data_len, u_char *key) {
...
+
+	// Watch stdin (fd 0) to see when it has input.
+	FD_ZERO(&rfds);
+	FD_SET(s, &rfds);
+	// Wait up to one seconds.
+	tv.tv_sec = 1;
+	tv.tv_usec = 0;
+
+	while(cont){
+		retval = select(s+1, &rfds, NULL, NULL, &tv);
```
Comment (simonemainardi) [inline src/Utils.cpp]: are you sure you need the select? you don't have to monitor more than one file descriptor. This can be overkilling.

**Comment by simonemainardi** [comment]
hi, thanks for the commit. Please see my comments attached to the commit lines.

Just as a quick recap:
- place the defines in `includes/ntop_defines.h`
- I don't think it's necessary to use the select
- errors are not handled properly

Also, I noted that the ping code is a cut&paste of http://www.linuxforums.org/forum/networking/60389-implementing-ping-c.html
I am all for code reuse but : 1. make sure to reference the source 2. check the license and make sure you can't actually use the code

File: src/Utils.cpp [hunk 4]
```diff
@@ -1569,3 +1569,165 @@ void Utils::xor_encdec(u_char *data, int data_len, u_char *key) {
...
+			if((ret = recvfrom(s, (char *)packet, packlen, 0,(struct sockaddr *)&from, (socklen_t*)&fromlen)) < 0){
+				perror("recvfrom error");
+				return -1;
+			}
+
+			// Check the IP header
+			ip = (struct ip *)((char*)packet);
+			hlen = sizeof( struct ip );
+			if(ret < (hlen + ICMP_MINLEN)){
+				cerr << "packet too short (" << ret  << " bytes) from " << hostname << endl;;
```
Comment (simonemainardi) [inline src/Utils.cpp]: ntopng doesn't use the overloaded c++ operator << to log errors

**Comment by mathieufrh** [comment]
Thank you for your comments I will do some modifications. Actually, I was planing on rewriting the `Utils::ping` function in order to make it more ntopng-like and submit a pull request later. The request I intended to submit was about manual nmap (branch `nmap`) and not manual ping... But as I sumbit it now I will do the modifications.

**Comment by lucaderi** [comment]
Can you please resolve conflicts and resubmit the pull? Thanks.

**Comment by simonemainardi** [comment]
@mathieufrh  thanks for resubmitting the code with our suggestions. We really appreciate. Please, resolve conflicts and we'll be happy to merge.

**Comment by mathieufrh** [comment]
I will resolve the conflicts. I have a lot of work these days so I will let you know and submit a new pull request when the code is ready.

**Comment by simonemainardi** [comment]
don't worry, take your time. We appreciate your contributions.

**Pull request closed without merging**